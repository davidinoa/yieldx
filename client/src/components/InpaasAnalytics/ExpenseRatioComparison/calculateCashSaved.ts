export default function calculateCashSaved(
  difference: number,
  portfolioMarketValue: number
) {
  return (difference / 100) * portfolioMarketValue;
}
