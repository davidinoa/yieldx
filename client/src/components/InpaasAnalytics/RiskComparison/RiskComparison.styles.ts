import styled from "styled-components";

export const ComparisonRow = styled.div`
  .comparison-label {
    font-family: var(--font-secondary);
    font-weight: bold;
    font-size: 11px;
    line-height: 13px;
    letter-spacing: 0.2em;
    text-transform: uppercase;
    color: var(--subtext-3);
  }

  .horizontal-separator {
    height: 2px;
    width: 80%;
    background: var(--border-new);
  }
  .vertical-separator {
    height: 80%;
    width: 2px;
    background: var(--border-new);
  }
  .comparison-value-container {
    margin: auto auto 0 auto;
  }
`;
