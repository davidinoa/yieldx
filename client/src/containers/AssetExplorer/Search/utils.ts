interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
