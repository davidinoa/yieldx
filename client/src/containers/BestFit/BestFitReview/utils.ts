import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
