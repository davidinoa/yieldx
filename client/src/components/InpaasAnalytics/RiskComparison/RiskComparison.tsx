import React from "react";
import { Comparison } from "@bondhouse/income-portfolios";
import PortfolioMetricCard from "components/PortfolioMetricCard/PortfolioMetricCard";
import ParentSize from "@visx/responsive/lib/components/ParentSize";
import CardValue from "components/PortfolioMetricCard/CardValue/CardValue";
import ComparisonBarChart from "components/ComparisonBarChart/ComparisonBarChart";
import { ComparisonRow } from "components/InpaasAnalytics/RiskComparison/RiskComparison.styles";
import { ToolTipContent } from "assets/content";
import { DataVal } from "components/ComparisonBarChart/util/dataIsEqual";
import { Spinner } from "elements/Spinner/Spinner";

interface Props {
  comparison?: Comparison;
}

function RiskComparison({ comparison }: Props) {
  const barData: [DataVal, DataVal] = [
    { label: "inpaas", value: comparison?.inpaas ?? 0 },
    {
      label: "average",
      value: comparison?.average ?? 0
    }
  ];
  return (
    <PortfolioMetricCard
      label="Risk comparison"
      xs={12}
      tooltipContent={ToolTipContent.riskComparison}
    >
      {comparison ? (
        <ComparisonRow className="row">
          <div className="col-7 d-flex flex-column justify-content-between">
            <div className="comparison-value-container">
              <CardValue
                value={{
                  number: comparison.percentageDifference ?? 0,
                  format: "%"
                }}
              />
              <span className="comparison-label">LESS RISK</span>
            </div>
            <div className="horizontal-separator m-auto" />
            <div className="comparison-value-container mt-0 mb-auto">
              <CardValue
                value={{
                  number: comparison.difference
                    ? comparison.difference * 100
                    : 0,
                  format: "bps"
                }}
              />
              <span className="comparison-label">REDUCTION</span>
            </div>
          </div>
          <div className="col-1 d-flex justify-content-center align-content-center">
            <div className="vertical-separator m-auto" />
          </div>
          <div className="col-16">
            <ParentSize>
              {({ width, height }) => (
                <ComparisonBarChart
                  width={width}
                  height={height}
                  data={barData}
                  maxValue={45}
                />
              )}
            </ParentSize>
          </div>
        </ComparisonRow>
      ) : (
        <ComparisonRow className="row">
          <Spinner />
        </ComparisonRow>
      )}
    </PortfolioMetricCard>
  );
}

export default RiskComparison;
