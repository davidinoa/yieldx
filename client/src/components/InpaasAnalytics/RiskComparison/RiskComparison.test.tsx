import React from "react";
import { render, screen } from "test/utils";
import { mockComparisons } from "components/InpaasAnalytics/ExpenseRatioComparison/calculateCashSaved.test";
import RiskComparison from "components/InpaasAnalytics/RiskComparison/RiskComparison";

describe("RiskComparison", () => {
  it("Should match snapshot", () => {
    const { asFragment } = render(
      <RiskComparison
        comparison={mockComparisons.find(it => it.comparisonType === "Risk")}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("Should render correct risk reduction percentage", () => {
    render(
      <RiskComparison
        comparison={mockComparisons.find(it => it.comparisonType === "Risk")}
      />
    );
    expect(screen.queryByText("78.57")).toBeInTheDocument();
  });

  it("Should render correct risk reduction bps", () => {
    render(
      <RiskComparison
        comparison={mockComparisons.find(it => it.comparisonType === "Risk")}
      />
    );
    expect(screen.queryByText("836")).toBeInTheDocument();
  });
});
