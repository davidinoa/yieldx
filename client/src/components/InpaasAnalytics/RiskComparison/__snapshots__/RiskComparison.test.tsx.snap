// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RiskComparison Should match snapshot 1`] = `
<DocumentFragment>
  .c2 > svg {
  margin: -0.5rem 0 0 0.35rem;
  vertical-align: middle;
  width: 0.85rem;
  height: 0.85rem;
}

.c0 {
  width: 100%;
  height: 100%;
  background: var(--dark);
  border: 1px solid var(--border-new);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: stretch;
  -webkit-justify-content: stretch;
  -ms-flex-pack: stretch;
  justify-content: stretch;
  border-radius: 0.5rem;
  padding: 1rem 1.15rem;
  min-height: 6rem;
}

.c0 > * {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.c1 {
  font-weight: bold;
  line-height: normal;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  text-transform: uppercase;
  color: var(--lightest);
  position: relative;
  font-size: 0.65rem;
  padding: 0.25rem 0;
  margin-bottom: 0.25rem;
  max-height: 2rem;
}

.c1 > svg {
  margin: -0.5rem 0 0 0.35rem;
  vertical-align: middle;
  width: 0.85rem;
  height: 0.85rem;
}

.c4 {
  line-height: normal;
  font-size: 1.5rem;
}

.c4 sup {
  font-size: 1rem;
  padding: 0.25rem;
}

.c4 span {
  color: var(--subtext-3);
  font-size: 12px;
  line-height: 15px;
}

.c4 .allocation-separator {
  height: 23px;
  width: 2px;
  background: var(--border-new);
}

.c3 .comparison-label {
  font-family: var(--font-secondary);
  font-weight: bold;
  font-size: 11px;
  line-height: 13px;
  -webkit-letter-spacing: 0.2em;
  -moz-letter-spacing: 0.2em;
  -ms-letter-spacing: 0.2em;
  letter-spacing: 0.2em;
  text-transform: uppercase;
  color: var(--subtext-3);
}

.c3 .horizontal-separator {
  height: 2px;
  width: 80%;
  background: var(--border-new);
}

.c3 .vertical-separator {
  height: 80%;
  width: 2px;
  background: var(--border-new);
}

.c3 .comparison-value-container {
  margin: auto auto 0 auto;
}

@media (min-width:1600px) {
  .c2 > svg {
    width: 1rem;
    height: 1rem;
  }
}

@media (min-width:1600px) {
  .c0 {
    border-radius: 0.625rem;
    padding: 1rem 1.65rem 1.65rem 1.65rem;
  }
}

@media (min-width:1600px) {
  .c1 {
    font-size: 0.75rem;
    margin-bottom: 0.5rem;
  }
}

@media (min-width:1600px) {
  .c1 > svg {
    width: 1rem;
    height: 1rem;
  }
}

@media (min-width:1600px) {
  .c4 {
    font-size: 1.95rem;
  }
}

@media (min-width:1600px) {
  .c4 sup {
    font-size: 1.15rem;
  }
}

<div
    class="mb-1 col-12"
    data-testid="Portfolio-Metric-Card"
  >
    <div
      class="c0"
    >
      <div
        class="c1 d-flex align-items-center"
      >
        Risk comparison
        <i
          class="c2"
        >
          <svg>
            info.svg
          </svg>
        </i>
      </div>
      <div
        class="c3 row"
      >
        <div
          class="col-7 d-flex flex-column justify-content-between"
        >
          <div
            class="comparison-value-container"
          >
            <div
              class="c4"
            >
              78.57
              <sup>
                %
              </sup>
            </div>
            <span
              class="comparison-label"
            >
              LESS RISK
            </span>
          </div>
          <div
            class="horizontal-separator m-auto"
          />
          <div
            class="comparison-value-container mt-0 mb-auto"
          >
            <div
              class="c4"
            >
              836
              <sup>
                bps
              </sup>
            </div>
            <span
              class="comparison-label"
            >
              REDUCTION
            </span>
          </div>
        </div>
        <div
          class="col-1 d-flex justify-content-center align-content-center"
        >
          <div
            class="vertical-separator m-auto"
          />
        </div>
        <div
          class="col-16"
        >
          <div
            style="width: 100%; height: 100%;"
          >
            <div
              style="width: 0px; height: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
